### ISSUE TITLE
Interface unification check does not consider tuples

### ISSUE THREAD

**Issue by CaptainHayashi** [dotnet/roslyn#12448]
**Version Used**: `master` 3466e42

**Steps to Reproduce**:

1. Build and launch a `VisualStudioSetup` instance for above version
2. Create a solution with the `dotnet-core` version of `System.ValueTuple` installed as a NuGet package
3. Add the following code to a C# file:

```csharp
    interface IBreak<A>
    {
        string Stringify(A a);
    }


    class Broken<A> : IBreak<(int, int)>, IBreak<(A, A)>
    {
        string IBreak<(int, int)>.Stringify((int, int) a)
        {
            return "foo";
        }
        string IBreak<(A, A)>.Stringify((A, A) a)
        {
            return "bar";
        }
    }
```

**Expected Behavior**:

The definition of `Broken` should (I think) cause an error similar to

`CS0695	'Broken<A>' cannot implement both 'IBreak<(int, int)>' and 'IBreak<(A, A)>' because they may unify for some type parameter substitutions`

because I could instantiate this with `Broken<int>` and have ambiguity between the two implementations of `Stringify`.  If I substitute `Tuple<X, Y>` for `(X, Y)` in the above, I get `CS0695` as expected.

**Actual Behavior**:

No such error is reported.  On trying to instantiate `Broken<int>` to trigger the above, I instead get `CS0648  'Broken2<int> is a type not supported by the language'`.

Having poked around the unification code (`Microsoft.CodeAnalysis.CSharp.TypeUnification`), there has been no change to this yet to support tuples.

This was mentioned in #10945 as a possible work-item, am posting to give evidence that this does seem to need updating.

**Comment by jcouv** [comment]
@jaredpar @gafter @VSadov I think this should be in RC. Let me know if otherwise.

**Comment by jcouv** [comment]
@CaptainHayashi Thanks for reporting this (great bug description). I expect to investigate this in this sprint (next 4 weeks).

**Comment by jcouv** [comment]
Fixed with https://github.com/dotnet/roslyn/pull/12689

**Comment by CaptainHayashi** [comment]
@jcouv Thanks!  This seems to fix the problem above.

I don't have a minimal working example for this yet (outside some experimental work in a fork that doesn't really apply to proper Roslyn), and I'm unsure if the problem is just theoretical, but the 'occurs' check seems to be ignoring tuples at the moment, eg. `TypeUnification.Contains((T1, T2), T1)` would return false (should be true).

Currently the generic case (eg `System.ValueTuple<T1, T2>`, `T1`) is handled in `Contains` by traversing every containing type of `System.ValueTuple<T1, T2>` and checking every type argument recursively with `Contains`.  Should the same be getting done for tuple types (eg converting to underlying type and then using this rule)?

**Comment by jcouv** [comment]
@CaptainHayashi 
I see that `TypeUnification.Contains` doesn't not handle tuples. I'm not sure if that is significant, but it might.
Let me try to find a scenario to hit that.

**Comment by jcouv** [comment]
@CaptainHayashi
I confirmed there is indeed a bug there as you suspected.
The code below will trigger an explosion (specifically an `OutOfMemoryException`) in `TypeUnification.CanUnify`, because `CanUnifyHelper` will return true and the assert below will attempt to verify that the substitution indeed yields matching types. The substitution map is `T -> Derived<(T, T)>` and the code for type substitution is aware of tuples and so will try to do that...

```C#
public interface I0<T1> { }
public class Derived<T> : I0<Derived<(T, T)>>, I0<T> { }
```

It only takes a few steps to get at something like this:
`NamedType I0<Derived<(Derived<(Derived<(Derived<(Derived<(T, T)>, Derived<(T, T)>)>, Derived<(Derived<(T, T)>, Derived<(T, T)>)>)>, Derived<(Derived<(Derived<(T, T)>, Derived<(T, T)>)>, Derived<(Derived<(T, T)>, Derived<(T, T)>)>)>)>, Derived<(Derived<(Derived<(Derived<(T, T)>, Derived<(T, T)>)>, Derived<(Derived<(T, T)>, Derived<(T, T)>)>)>, Derived<(Derived<(Derived<(T, T)>, Derived<(T, T)>)>, Derived<(Derived<(T, T)>, Derived<(T, T)>)>)>)>)>>`


Also, your comment prompted me to systematically hunt down other places where tuple handling is missing (by reviewing special handling of arrays, looking for usages of `SymbolKind.ArrayType`; arrays being very similar to tuples in many ways).
Found a dozen gaps this way (see https://github.com/dotnet/roslyn/issues/12916), including one that is pretty serious (accessibility check). 

Thanks again! :hand:

**Comment by jcouv** [comment]
The remaining issue is tracked by #12916.
